File: upwork-intelligence/merge-raw.py
#!/usr/bin/env python3
"""Merge raw_batches/*.json into search-raw.json for process-run.mjs."""
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent
KEYWORDS = json.loads((ROOT / "keywords.json").read_text())
BATCH_DIR = ROOT / "raw_batches"
BATCH_DIR.mkdir(exist_ok=True)

searches = []
errors = []
done = set()

for p in sorted(BATCH_DIR.glob("*.json")):
    data = json.loads(p.read_text())
    kw = data.get("keyword")
    if not kw:
        continue
    done.add(kw)
    searches.append(
        {
            "keyword": kw,
            "group": data.get("group"),
            "jobs": data.get("jobs", []),
        }
    )

for item in KEYWORDS:
    kw = item["keyword"]
    if kw not in done:
        errors.append(kw)

out = {"searches": searches, "errors": errors}
(ROOT / "search-raw.json").write_text(json.dumps(out))
print(json.dumps({"searches": len(searches), "errors": len(errors)}, indent=2))

File: upwork-intelligence/merge-raw.mjs
#!/usr/bin/env node
import fs from 'fs';
import path from 'path';

const dir = path.dirname(new URL(import.meta.url).pathname);
const batchDir = path.join(dir, 'raw_batches');
const keywords = JSON.parse(fs.readFileSync(path.join(dir, 'keywords.json'), 'utf8'));

const searches = [];
const errors = [];

if (fs.existsSync(batchDir)) {
  for (const f of fs.readdirSync(batchDir).filter((x) => x.endsWith('.json'))) {
    const batch = JSON.parse(fs.readFileSync(path.join(batchDir, f), 'utf8'));
    if (Array.isArray(batch)) {
      for (const entry of batch) searches.push(entry);
    } else if (batch.searches) {
      searches.push(...batch.searches);
    }
  }
}

const byKw = new Map();
for (const s of searches) {
  if (s.keyword) byKw.set(s.keyword, s);
}

const merged = [];
for (const kw of keywords) {
  const ex = byKw.get(kw);
  if (ex) merged.push(ex);
  else errors.push(kw);
}

const payload = {
  keywordsAttempted: keywords.length,
  keywordsCompleted: merged.filter((s) => !s.error).length,
  searches: merged,
};
const out = path.join(dir, '.run-payload.json');
fs.writeFileSync(out, JSON.stringify(payload));
console.log(JSON.stringify({ out, merged: merged.length, missing: errors.length, errors }));
